gitlab/resource_deploykey.go

```diff
@@ -0,0 +1,197 @@
+	}
+
+	return dk.c.c.DeleteKey(ctx, getRepoPath(dk.c.ref), dk.k.ID)
+}
+
+// Reconcile makes sure the desired state in this object (called "req" here) becomes
+// the actual state in the backing Git provider.
+//
+// If req doesn't exist under the hood, it is created (actionTaken == true).
+// If req doesn't equal the actual state, the resource will be updated (actionTaken == true).
+// If req is already the actual state, this is a no-op (actionTaken == false).
+//
+// The internal API object will be overridden with the received server data if actionTaken == true.
+func (dk *deployKey) Reconcile(ctx context.Context) (bool, error) {
+	actual, err := dk.c.get(ctx, dk.k.Title)
+	if err != nil {
+		// Create if not found
+		if errors.Is(err, gitprovider.ErrNotFound) {
+			return true, dk.createIntoSelf(ctx)
+		}
+
+		// Unexpected path, Get should succeed or return NotFound
+		return false, err
+	}
+
+	// Use wrappers here to extract the "spec" part of the object for comparison
+	desiredSpec := newGitlabKeySpec(&dk.k)
+	actualSpec := newGitlabKeySpec(&actual.k)
+
+	// If the desired matches the actual state, do nothing
+	if desiredSpec.Equals(actualSpec) {
+		return false, nil
+	}
+	// If desired and actual state mis-match, update
+	return true, dk.Update(ctx)
+}
+
+func (dk *deployKey) createIntoSelf(ctx context.Context) error {
+	// POST /repos/{owner}/{repo}/keys
+	apiObj, err := dk.c.c.CreateKey(ctx, getRepoPath(dk.c.ref), &dk.k)
+	if err != nil {
+		return err
+	}
+	dk.k = *apiObj
+	return nil
+}
+
+func validateDeployKeyAPI(apiObj *gitlab.DeployKey) error {
+	return validateAPIObject("GitLab.Key", func(validator validation.Validator) {
+		if apiObj.Title == "" {
+			validator.Required("Title")
+		}
+		if apiObj.Key == "" {
+			validator.Required("Key")
+		}
+	})
+}
+
+func deployKeyFromAPI(apiObj *gitlab.DeployKey) gitprovider.DeployKeyInfo {
+	return gitprovider.DeployKeyInfo{
+		Name: apiObj.Title,
+		Key:  []byte(apiObj.Key),
+	}
+}
+
+func deployKeyToAPI(info *gitprovider.DeployKeyInfo) *gitlab.DeployKey {
+	k := &gitlab.DeployKey{}
+	deployKeyInfoToAPIObj(info, k)
+	return k
+}
+
+func deployKeyInfoToAPIObj(info *gitprovider.DeployKeyInfo, apiObj *gitlab.DeployKey) {
+	// Required fields, we assume info is validated, and hence these are set
+	apiObj.Title = info.Name
+	apiObj.Key = string(info.Key)
+	// optional fields
+	derefedBool := false
+	if info.ReadOnly != nil {
+		if *info.ReadOnly {
+			apiObj.CanPush = &derefedBool
+		} else {
+			derefedBool = true
+			apiObj.CanPush = &derefedBool
+		}
+	}
+}
+
+// This function copies over the fields that are part of create request of a deploy
+// i.e. the desired spec of the deploy key. This allows us to separate "spec" from "status" fields.
+func newGitlabKeySpec(key *gitlab.DeployKey) *gitlabKeySpec {
+	return &gitlabKeySpec{
+		&gitlab.Key{
+			// Create-specific parameters
+			Title: key.Title,
+			Key:   key.Key,
```

Comment:
Missing ReadOnly/CanWrite here, important at all? I can't quite tell..

Reply:
Added, thanks for the catch :+1: